protocol/test-serialization/src/graphql_blocks.rs

```diff
@@ -0,0 +1,93 @@
+
+#[cfg(test)]
+mod tests {
+    use mina_rs_base::types::ExternalTransition;
+    use mina_rs_base::JsonSerializationType;
+    use serde::{Deserialize, Serialize};
+    use serde_json::Error;
+
+    // To read query response from graphql api (https://graphql.minaexplorer.com/)
+    #[derive(Debug, Serialize, Deserialize)]
+    struct QueryResponse {
+        data: Blocks,
+    }
+    #[derive(Debug, Serialize, Deserialize)]
+    struct Blocks {
+        blocks: Vec<BlockInfo>,
+    }
+    #[derive(Debug, Serialize, Deserialize)]
+    struct BlockInfo {
+        #[serde(rename = "blockHeight")]
+        block_height: usize,
+        #[serde(rename = "stateHash")]
+        state_hash: String,
+    }
+
+    /// Ensures coverage of JSON serde by testing n latest blocks from mainnet
+    #[tokio::test]
+    async fn check_latest_blocks_compatibility() {
+        let mut check_status = true;
+        // Fetch n blocks info from graphql api
+        let blocks = fetch_block_info().await.unwrap();
+        for block in blocks.data.blocks {
+            // Retrieve block json from mainnet json dump
+            let block_json = match retrieve_block_json("mainnet", &block).await {
+                Ok(val) => val,
+                Err(err) => {
+                    println!("Failed to retrieve block json: {}", err);
+                    continue;
+                }
+            };
+            match read_block_json(block_json) {
+                Ok(_et) => {} // Block read successfully
+                Err(err) => {
+                    check_status = false;
+                    println!(
+                        "Fix Me({}-{}): {}",
+                        block.block_height, block.state_hash, err
+                    );
+                    continue;
+                }
+            }
+        }
+        assert!(check_status)
+    }
+
+    async fn fetch_block_info() -> Result<QueryResponse, reqwest::Error> {
+        let client = reqwest::Client::new();
+        let res = client
+            .post("https://graphql.minaexplorer.com/")
+            // Below query retrieves blockHeight and stateHash of 10 latest blocks
+            // To learn more about graphql query refer to https://docs.minaprotocol.com/en/developers/graphql-api 
+            .json(&serde_json::json!({
+                "query":"query Fetch {\n  blocks(limit: 10, sortBy: RECEIVEDTIME_DESC) {\n    blockHeight\n    stateHash\n  }\n}"
+            }))
+            .send()
+            .await?;
+        Ok(res.json::<QueryResponse>().await?)
+    }
+
+    async fn retrieve_block_json(
+        network: &str,
+        block: &BlockInfo,
+    ) -> Result<serde_json::Value, reqwest::Error> {
+        let block_url = format!(
+            "https://storage.googleapis.com/mina_network_block_data/{}-{}-{}.json",
+            network, block.block_height, block.state_hash
+        );
+        let mut block_json: serde_json::Value = reqwest::get(block_url).await?.json().await?;
```

Comment:
Retry here? I would expect this test being flaky, not sure if it's good idea to run this in every CI job, @creativcoder any ideas?

Reply:
Also IMHO 10 is too small, I believe the goal is to detect incompatible blocks but not blocking CI when detected

Reply:
Yeah, this test is only ideal to run the manual way on need basis. I would move it out of CI (maybe under scripts/ as test_blocks_compat.rs).

Reply:
I created a new dir `tools` and added `blocks-crawler`
also excluded it from our current workspace
please review it once again